Components architecture and examples. (#22)
ros-controls/roadmap: design_drafts/components_architecture_and_urdf_examples.md

# ROS2 Control Components Architecture and URDF-Description Examples

## Architecture

![ros2_control Components Achitecture][ros2_control_arch_resource_manager]

[ros2_control_arch_resource_manager]: images/components_architecture.png "ROS2 Control - Components Architecture"

## URDF Examples

ros2_control implements following robot/robot-cell architectures:

1. Industrial Robots with only one interface
2. Industrial Robots with multiple interfaces (can not be written at those at the same time)
3. Robots with multiple interfaces used at the same time
4. Industrial Robots with a sensor integrated into the robot's control box
5. Industrial Robots with a sensor connected to ROS computer
6. Modular Robots with separate communication to each actuator
7. Modular Robots with actuators not providing states and additional sensors (simple Transmissions)
8. Modular Robots with separate communication to each "actuator" with multi joints (Transmission Example) - (system component is used)
9. Sensor only
10. Actuator Only

> 

Reviewer: ```suggestion

Note:
  * Everything within the `<classType>` tag is implemented as a plugin.
  * The examples below have some `<param>` tags defined for each plugin which are primarily for demonstration, not part of a pre-defined xml schema. Components can define their own parameters.
```

#### 1. Industrial Robots with only one interface
  * the communication is done using proprietary API to connect the robots control box
  * Data for all joints is exchanged in batch (at once)
  * Examples: KUKA RSI

```xml
  <ros2_control name="2DOF_System_Robot_Position_Only" type="system">
    <hardware>
      <classType>ros2_control_demo_hardware/2DOF_System_Hardware_Position_Only</classType>
      <param name="example_param_write_for_sec">2</param>
      <param name="example_param_read_for_sec">2</param>
    </hardware>
    <joint name="joint1">
      <classType>ros2_control_components/PositionJoint</classType>
      <param name="min_values">-1</param>
      <param name="max_values">1</param>
    </joint>
    <joint name="joint2">
      <classType>ros2_control_components/PositionJoint</classType>
      <param name="min_values">-1</param>
      <param name="max_values">1</param>
    </joint>
  </ros2_control>
```

Note:
  * `ros2_control_components/PositionJoint`type has implicitly:
    ```xml
      <commandInterfaceType>position</commandInterfaceType>
      <stateInterfaceType>position</stateInterfaceType>
    ```

> #### 2. Industrial Robots with multiple interfaces (can not be written at those at the same time)

Reviewer: ```suggestion
#### 2. Industrial Robots with multiple interfaces (can not be written at the same time)
```

>   * the communication is done using proprietary API to connect the robots control box

Reviewer: ```suggestion
  * the communication is done using proprietary API to communicate with robot control box
```

      <commandInterfaceType>effort</commandInterfaceType>
      <stateInterfaceType>position</stateInterfaceType>
      <stateInterfaceType>velocity</stateInterfaceType>
      <stateInterfaceType>effort</stateInterfaceType>
      <param name="min_position_value">-1</param>
      <param name="max_position_value">1</param>
      <param name="min_velocity_value">-1</param>
      <param name="max_velocity_value">1</param>
      <param name="min_effort_value">-0.5</param>
      <param name="max_effort_value">0.5</param>
    </joint>
    <joint name="joint2">
      <classType>ros2_control_components/MultiInterfaceJoint</classType>
      <commandInterfaceType>position</commandInterfaceType>
      <stateInterfaceType>position</stateInterfaceType>
      <stateInterfaceType>velocity</stateInterfaceType>
      <stateInterfaceType>effort</stateInterfaceType>
      <param name="min_position_value">-1</param>
      <param name="max_position_value">1</param>
      <param name="min_velocity_value">-1</param>
      <param name="max_velocity_value">1</param>
      <param name="min_effort_value">-0.5</param>
      <param name="max_effort_value">0.5</param>
    </joint>
  </ros2_control>
```

Note:

>   * For `joint2` the "velocity" and "effort" command interfaces are intentionally left out to show another often use-case

Reviewer: ```suggestion
  * For `joint2` the "velocity" and "effort" command interfaces are intentionally left out to show another common use-case
```

      <classType>ros2_control_components/PositionJoint</classType>
      <stateInterfaceType>position</stateInterfaceType>
      <param name="min_values">${-PI}</param>
      <param name="max_values">${PI}</param>
    </joint>
  </ros2_control>
  <ros2_control name="2DOF_System_Robot_Position_Sensor_joint2" type="sensor">
    <hardware>
      <classType>ros2_control_demo_hardware/Position_Sensor_Hardware</classType>
      <param name="example_param_read_for_sec">2</param>
    </hardware>
    <joint name="joint2">
      <classType>ros2_control_components/PositionJoint</classType>
      <stateInterfaceType>position</stateInterfaceType>
      <param name="min_values">${-PI}</param>
      <param name="max_values">${PI}</param>
    </joint>
  </ros2_control>
```
Note:
  * since there is only one keyword `commandInterfaceType` or `stateInterfaceType` given to `ros2_control_components/PositionJoint`type the other one is ignored for the hardware

#### 8. Modular Robots with separate communication to each "actuator" with multi joints (Transmission Example) - (system component is used)
  * the communication is done on actuator level using proprietary or standardized API (e.g., canopen_402)

>   * Data for all actuator is exchanged separately from each other

Reviewer: ```suggestion
  * Data for all actuators is exchanged separately from each other
```

>   * Examples: Wrist of a humanoid robot

Reviewer: ```suggestion
  * There is a many-to-many connection between joint and actuator values resolved by a transmission
  * Examples: Wrist of a humanoid robot
```

        <mechanicalReduction_output1>3.1</mechanicalReduction_output1>
        <mechanicalReduction_output2>1.4</mechanicalReduction_output2>
      </joint>
    </transmission>
  </ros2_control>
```

#### 9. Sensor only
  * the communication is done on a sensor level using proprietary or standardized API
  * Examples: Camera, ForceTorqueSensor, Distance Sensors, IMU, etc.

```xml
  <ros2_control name="Camera_with_IMU"  type="sensor">
    <hardware>
      <classType>ros2_control_demo_hardware/CameraWithIMU_Sensor</classType>
      <param name="example_param_read_for_sec">2</param>
    </hardware>
    <sensor name="sensor1">
      <classType>ros2_control_components/IMUSensor</classType>
      <interfaceType>velocity</interfaceType>
      <interfaceType>acceleration</interfaceType>
      <param name="min_values">-1000</param>
      <param name="max_values">1000</param>
    </joint>
    <joint name="sensor2">
      <classType>ros2_control_components/2DImageSensor</classType>
      <interfaceType>image</interfaceType>
      <param name="min_values">0</param>
      <param name="max_values">255</param>
    </joint>
  </ros2_control>
```

#### 10. Actuator Only

>   * the communication is done on actuator level using proprietary or standardized API

Reviewer: ```suggestion
  * the communication is done on actuator level using proprietary or standardized API
  * There may be a mechanical reduction or other type of 1-1 mapping between joint and actuator resolved by a transmission
```

# ROS2 Control Components Architecture and URDF-Description Examples

## Architecture

![ros2_control Components Achitecture][ros2_control_arch_resource_manager]

> [ros2_control_arch_resource_manager]: images/components_architecture.png "ROS2 Control - Components Architecture"

Author: > Not sure how to comment on the image, so I'll do it here:
> I think it's an implementation detail to mention "ActuatorHardwareInterface" and would remove that row of boxes with "template to implement ..."

I wanted to have this implementation detail in it. The image is something between implementation UML and architecture and I hope for a user would be clear where to starts its implementation.

> As for wording, I know it's a constant bike shedding, but I don't see the point to add "Hardware" to "ActuatorHardware". Same for sensors and systems.

I did it mainly to make a difference between "Joint" and "Sensor" on the right side, and those implementations of interfacing the hardware directly. This wording is in line with ROS1 ros_control with `RobotHardware` and `RobotHardwareInterface`.

> However, what this document lacks in my opinion is a general introduction/motivation and explanation of wording. Basically saying what's meant with "actuator", what's a system etc.

I agree with you. My intention of current state of PR is to clarify the URDF structures, so we can start implement this in the main repository.

## URDF Examples

> ros2_control implementation examples are presented for the following robot/robot-cell architectures:

Author: What is the proper English term for a fenced area where robot is, usually with other machines like feeders, doing his thing? [image](https://www.technische-logistik.net/sites/default/files/styles/large/public/bilder/MRZ_2.1.jpg?itok=Nx750pBE)

In German is called "RoboterZellle". Therefore, I thought "cell" is a good term.

> ros2_control implementation examples are presented for the following robot/robot-cell architectures:

Reviewer: Caged robots is the common term in English I believe

1. Industrial Robots with only one interface
2. Industrial Robots with multiple interfaces (can not be written at the same time)
3. Robots with multiple interfaces used at the same time
4. Industrial Robots with a sensor integrated into the robot's control box
5. Industrial Robots with a sensor connected to ROS computer
6. Modular Robots with separate communication to each actuator
7. Modular Robots with actuators not providing states and additional sensors (simple Transmissions)
8. Modular Robots with separate communication to each "actuator" with multi joints (Transmission Example) - (system component is used)
9. Sensor only
10. Actuator Only

Note:
  * Everything within the `<classType>` tag is implemented as a plugin.

>   * The examples below have some `<param>` tags defined for each plugin which are primarily for demonstration, not part of a pre-defined xml schema. Components can define their own parameters.

Author: I reword this. Is it better now?

> But we have to maybe clarify how these parameters are being used. 

This would be done in the example implementation.

#### 1. Industrial Robots with only one interface
  * the communication is done using proprietary API to communicate with robot control box
  * Data for all joints is exchanged in batch (at once)
  * Examples: KUKA RSI

```xml
  <ros2_control name="2DOF_System_Robot_Position_Only" type="system">
    <hardware>
      <classType>ros2_control_demo_hardware/2DOF_System_Hardware_Position_Only</classType>
      <param name="example_param_write_for_sec">2</param>
      <param name="example_param_read_for_sec">2</param>
    </hardware>

>     <joint name="joint1">

Author: > this joint is purely virtual and abstract to any hardware specific implementation

I do not agree. On this level, in the `RessourceManager` all joints are actually the real joints of the hardware. First when we go into `ControllerManager` virtual joints are used. (This is also shown, not in full detail, in the image)

>     <joint name="joint1">

Reviewer 2: Then I still don't understand what's happening with joints. In my understanding, the term joint here described something abstract which encapsulates common pattern like joint limits etc before these values are then finally passed to the system implementation.

>     <joint name="joint1">

Reviewer: There's the forward facing XML format we are trying to set up and the code that this will eventually tie into. I would 100% consider joint_handle an implementation detail nobody should care about when defining/reading a URDF. No matter how this is implemented, the URDF should never refer to handles, let's keep it at the abstract terms level.

      <classType>ros2_control_components/PositionJoint</classType>
      <param name="min_values">-1</param>
      <param name="max_values">1</param>
    </joint>
    <joint name="joint2">
      <classType>ros2_control_components/PositionJoint</classType>
      <param name="min_values">-1</param>
      <param name="max_values">1</param>
    </joint>
  </ros2_control>
```

Note:
  * `ros2_control_components/PositionJoint`type has implicitly:
    ```xml
      <commandInterfaceType>position</commandInterfaceType>
      <stateInterfaceType>position</stateInterfaceType>

>     ```

Author: We should add some general description at the beginning.

> #### 2. Industrial Robots with multiple interfaces (can not be written at the same time)

Author: What about:

"ControlState" and "ControlCommand"?

@bmagyar

> #### 2. Industrial Robots with multiple interfaces (can not be written at the same time)

Reviewer: We shouldn't be picky about overloading general terms after what's been done with  `DynamicJointState`

Let's call it `joint interfaces` as our `DynamicJointState` message reflects. I think this is fine

  * the communication is done using proprietary API to communicate with robot control box
  * Data for all joints is exchanged in batch (at once)
  * Examples: KUKA FRI, ABB Yummy, Schunk LWA4p, etc.

```xml
  <ros2_control name="2DOF_System_Robot_MultiInterface" type="system">
    <hardware>
      <classType>ros2_control_demo_hardware/2DOF_System_Hardware_MultiInterface</classType>
      <param name="example_param_write_for_sec">2</param>
      <param name="example_param_read_for_sec">2</param>
    </hardware>
    <joint name="joint1">
      <classType>ros2_control_components/MultiInterfaceJoint</classType>

>       <commandInterfaceType>position</commandInterfaceType>

Reviewer 2: Do we have to make sure at this point that the modes are not mixed up between the joints? Or does it make sense to control joint1 in position and control2 in effort?

Author: IMO, we do not have to put limitation on this. I am actually actively using this: on my robot there are six joints, three of them are position-controlled and three velocity-controlled.

      <commandInterfaceType>effort</commandInterfaceType>
      <stateInterfaceType>position</stateInterfaceType>
      <stateInterfaceType>velocity</stateInterfaceType>
      <stateInterfaceType>effort</stateInterfaceType>
      <param name="min_position_value">-1</param>
      <param name="max_position_value">1</param>
      <param name="min_velocity_value">-1</param>
      <param name="max_velocity_value">1</param>
      <param name="min_effort_value">-0.5</param>
      <param name="max_effort_value">0.5</param>
    </joint>
    <joint name="joint2">
      <classType>ros2_control_components/MultiInterfaceJoint</classType>
      <commandInterfaceType>position</commandInterfaceType>
      <stateInterfaceType>position</stateInterfaceType>
      <stateInterfaceType>velocity</stateInterfaceType>
      <stateInterfaceType>effort</stateInterfaceType>
      <param name="min_position_value">-1</param>
      <param name="max_position_value">1</param>
      <param name="min_velocity_value">-1</param>
      <param name="max_velocity_value">1</param>
      <param name="min_effort_value">-0.5</param>
      <param name="max_effort_value">0.5</param>
    </joint>
  </ros2_control>
```

Note:

>   * For `joint2` the "velocity" and "effort" command interfaces are intentionally left out to show another common use-case

Reviewer 2: make highlight this inline with a commented out tag.

Author: sorry, I don't understand what you mean...

Reviewer 2: sorry, incomplete sentence here. My idea was to have the velocity and effort tags in the example, but being commented out with a comment on top.

#### 3. Robots with multiple interfaces used at the same time
  * the communication is done using proprietary API or some standardized interface
  * Data for all joints is exchanged in batch (at once)
  * Examples: (some humanoid robots)?

```xml
  <ros2_control name="2DOF_System_Robot_MultiInterface_MultiWrite" type="system">
    <hardware>

>       <classType>ros2_control_demo_hardware/2DOF_System_Hardware_MultiInterface_MultiWrite</classType>

Reviewer 2: consider leaving out this use case

Author: You are right, there is no really a value in it. I will remove it but leave the use-case named. This will be cleaned when finalizing the document.

      <stateInterfaceType>effort</stateInterfaceType>
      <param name="min_position_value">-1</param>
      <param name="max_position_value">1</param>
      <param name="min_velocity_value">-1</param>
      <param name="max_velocity_value">1</param>
      <param name="min_effort_value">-0.5</param>
      <param name="max_effort_value">0.5</param>
    </joint>
    <joint name="joint2">
      <classType>ros2_control_components/MultiInterfaceJoint_MultiWrite</classType>
      <commandInterfaceType>position</commandInterfaceType>
      <commandInterfaceType>velocity</commandInterfaceType>
      <commandInterfaceType>effort</commandInterfaceType>
      <stateInterfaceType>position</stateInterfaceType>
      <stateInterfaceType>velocity</stateInterfaceType>
      <stateInterfaceType>effort</stateInterfaceType>
      <param name="min_position_value">-1</param>
      <param name="max_position_value">1</param>
      <param name="min_velocity_value">-1</param>
      <param name="max_velocity_value">1</param>
      <param name="min_effort_value">-0.5</param>
      <param name="max_effort_value">0.5</param>
    </joint>
  </ros2_control>
```

> #### 4. Industrial Robots with a sensor integrated into the robot's control box

Reviewer 2: Compared to use case number 1, what does change for me as a developer who's implementing the system's plugin?
Also, consider moving this example directly following number1

Reviewer 2: Describe what a control box is here

Author: will do

> #### 4. Industrial Robots with a sensor integrated into the robot's control box

Reviewer 2: ```suggestion
#### 4. Industrial Robots with integrated sensor
```

    <hardware>
      <classType>ros2_control_demo_hardware/2DOF_System_Hardware_Sensor</classType>
      <param name="example_param_write_for_sec">2</param>
      <param name="example_param_read_for_sec">2</param>
    </hardware>
    <joint name="joint1">
      <classType>ros2_control_components/PositionJoint</classType>
      <param name="can_read">True</param>
      <param name="min_values">-1</param>
      <param name="max_values">1</param>
    </joint>
    <joint name="joint2">
      <classType>ros2_control_components/PositionJoint</classType>
      <param name="can_read">True</param>
      <param name="min_values">-1</param>
      <param name="max_values">1</param>
    </joint>
    <sensor name="tcp_fts_sensor">
      <classType>ros2_control_components/ForceTorqueSensor</classType>
      <param name="frame_id">kuka_tcp</frame_id>
      <param name="lower_limits">-100</param>
      <param name="upper_limits">100</param>
    </sensor>
  </ros2_control>
```
Note:
  * `ros2_control_components/PositionJoint`type has implicitly:
    ```xml
      <commandInterfaceType>position</commandInterfaceType>
      <stateInterfaceType>position</stateInterfaceType>
    ```

> #### 5. Industrial Robots with a sensor connected to ROS computer

Reviewer 2: ```suggestion
#### 5. Industrial Robots with externally connected sensor
```

  * the communication is done using proprietary API
  * Data for all joints is exchanged at once
  * Sensor data are exchanged independently
  * Examples: KUKA RSI and FTS connected to ROS-PC

```xml
  <ros2_control name="2DOF_System_Robot_Position_Only_External_Sensor" type="system">
    <hardware>
      <classType>ros2_control_demo_hardware/2DOF_System_Hardware_Position_Only</classType>
      <param name="example_param_write_for_sec">2</param>
      <param name="example_param_read_for_sec">2</param>
    </hardware>
    <joint name="joint1">
      <classType>ros2_control_components/PositionJoint</classType>
      <param name="can_read">True</param>
      <param name="min_values">-1</param>
      <param name="max_values">1</param>
    </joint>
    <joint name="joint2">
      <classType>ros2_control_components/PositionJoint</classType>
      <param name="can_read">True</param>
      <param name="min_values">-1</param>
      <param name="max_values">1</param>
    </joint>
  </ros2_control>

>   <ros2_control name="2DOF_System_Robot_ForceTorqueSensor" type="sensor">

Reviewer 2: I think it deserves a bit more context or description that we have two separate `ros2_control` tags in here and thus treat the two components individually.

      <param name="min_values">-1</param>
      <param name="max_values">1</param>
    </joint>
  </ros2_control>
```

#### 7. Modular Robots with actuators not providing states and additional sensors (simple Transmissions)
  * the communication is done on actuator level using proprietary or standardized API (e.g., canopen_402, modbus, RS232, RS485)
  * Data for all actuators and sensors is exchanged separately from each other
  * Examples: Arduino-based-robots, custom robots

```xml
  <ros2_control name="2DOF_Modular_Robot_joint1"  type="actuator">
    <hardware>
      <classType>ros2_control_demo_hardware/Velocity_Actuator_Hadware</classType>
      <param name="example_param_write_for_sec">1.23</param>
      <param name="example_param_read_for_sec">3</param>
    </hardware>
    <joint name="joint1">
      <classType>ros2_control_components/VelocityJoint</classType>
      <commandInterfaceType>position</commandInterfaceType>
      <param name="min_values">-1</param>
      <param name="max_values">1</param>
    </joint>
    <transmission name="transmission1">
      <type>transmission_interface/SimpleTansmission</type>

>       <joint name="joint1">

Reviewer 2: I could see how "joint" becomes quite confusing here.

Reviewer 2: can we not just directly pass in all listed joints within the `ros2_control` tag to the transmission plugin? I mean it doesn't make much sense to specify "joint5" here if that joint is never being used within that component.

Author: This is true. For the actuators we can do this implicitly.

        <interfaceType>velocity</interfaceType>

>         <velocity_to_voltage>${1024/PI}</velocity_to_voltage>

Reviewer 2: ```suggestion
        <param name="velocity_to_voltage">${1024/PI}</param>
```

Author: Great catch!
